### DISCUSSION TITLE
How to flash tasmota to an esp8266 device with ArduinoOTA

### DISCUSSION THREAD

**Discussion by jeroenst** [arendst/Tasmota#12978] [Q&A] [answered]
I have a lot of esp8266 devices with my own firmware which includes arduinoota (https://www.arduino.cc/reference/en/libraries/arduinoota/).

Upgrading with my own firmware succeeds always without any problem.

Last time I tried to flash Tasmota ota to one of my devices but it didn't work and the device became bricked.

I had to remove the device, open it and flash it by serial before it worked again.

What is the best way to flash Tasmota ota to my devices running their own firmware at this moment without bricking them?

**Comment by ascillato** [comment]
> What is the best way to flash Tasmota ota to my devices running their own firmware at this moment without bricking them?

Both must have the same memory linker map in order to not scramble the SDK code and settings.

`board                       = esp8266_1M`

https://github.com/arendst/Tasmota/blob/d4f444a85af216ffc0cc1cb8c22fc92685a0b265/platformio.ini#L27-L31

**Reply by jeroenst** [reply]
@ascillato  So changing from esp8266_4M to esp8266_1M is not possible ota?
And so I can't use precompiled Tasmota binaries?

**Reply by barbudor** [reply]
You need to compare the map exactly
You may have to build intermediary Arduino Ota firmware to OTA between your current firmware and Tasmota
I would suggest that you experiment on a D1/Nodemcu to avoid bricking a installed not-accessible device

Tasmota default to 1mb to insure max interoperability with consumer devices (sonoff...) but there are options in Tasmota to explicitly target 4mb boards like when you want a file system. Not sure it can help, may be.
But generally indeed Ota between different maps is difficult (in general, not specific to tasmota or esp)

**Reply by jeroenst** [reply]
This is the build command I use for my own firmware:

esp8266com:esp8266:nodemcuv2:xtal=80,vt=flash,exception=disabled,ssl=all,eesz=4M,led=2,ip=lm2f,dbg=Disabled,lvl=None____,wipe=none
```
nodemcuv2.menu.eesz.4M=4MB (FS:none OTA:~1019KB)
nodemcuv2.menu.eesz.4M.build.flash_size=4M
nodemcuv2.menu.eesz.4M.build.flash_size_bytes=0x400000
nodemcuv2.menu.eesz.4M.build.flash_ld=eagle.flash.4m.ld
nodemcuv2.menu.eesz.4M.build.spiffs_pagesize=256
nodemcuv2.menu.eesz.4M.upload.maximum_size=1044464
nodemcuv2.menu.eesz.4M.build.rfcal_addr=0x3FC000
```

For 1MB devices I use;

esp8266com:esp8266:esp8285:xtal=80,vt=flash,exception=disabled,ssl=all,ResetMethod=nodemcu,CrystalFreq=26,eesz=1M,led=2,ip=lm2f,dbg=Disabled,lvl=None____,wipe=none,baud=115200

```
esp8285.menu.eesz.1M=1MB (FS:none OTA:~502KB)
esp8285.menu.eesz.1M.build.flash_size=1M
esp8285.menu.eesz.1M.build.flash_size_bytes=0x100000
esp8285.menu.eesz.1M.build.flash_ld=eagle.flash.1m.ld
esp8285.menu.eesz.1M.build.spiffs_pagesize=256
esp8285.menu.eesz.1M.upload.maximum_size=1023984
esp8285.menu.eesz.1M.build.rfcal_addr=0xFC000
```

The device I bricked was a sonoff 4ch pro, which used the 1 MB image, I didn't try 4MB devices yet.

**Comment by arendst** [comment, marked as answer]
This discussion leads nowhere.

Installing tasmota using OTA on a non-tasmota device is not supported.

You should have used serial flash to upgrade to tasmota. End of discussion.

**Reply by jeroenst** [reply]
Ok? Thank you? This answer feels pretty rough to me. I was just trying to ask what I could do to flash tasmota on a esp8266 ota (like tuya devices) and wanted to learn why it doesn't work...  @barbudor did help me already without just closing my question, which i'm thankful for. Going over 30+ devices which are installed in walls and ceilings without ota is a hell of a job, that's why I'm wondering if there wasn't a way to avoid this. Shellies i.e. Can be flashed over the air without any problem.

**Comment by jeroenst** [comment]
Related Question: https://github.com/arendst/Tasmota/issues/5048

**Reply by Jason2866** [reply]
Your question is not closed, the project owner answered. The problem if you think it worked unexpected issues can occur. Tasmota is not designed to be initial loaded that way.
You can try to load https://github.com/Jason2866/ESP8266MinimalHTTPUpdater and flash Tasmota via this. No support! and as said before dont be surprised if Tasmota behaves wonky
#5048 is a different question, not related. Device was not flashed with ArduinoOTA.
About Tuya (convert) the designed the app to do it correctly. You can look into there code how to do. There where a lot of bricked Tuya devices on there way...

**Reply by jeroenst** [reply]
No problem if it's not possible but I try to understand what's causing this. 

I will try to give your link a shot on a test device.

When I succeed in flashing the latest Tasmota in the device and it works fine, is there still any difference?

Thank you

**Reply by Jason2866** [reply]
Yes, still a difference. Since flash is not erased where Tasmota stores the settings. So with the old left overs in flash everything can happen. Tuya solved by erasing this area in flash.

**Reply by jeroenst** [reply]
And a factory reset won't help for this?

**Reply by barbudor** [reply]
If Tasmota starts and you can issue a [`reset 2`](https://tasmota.github.io/docs/Commands/#reset) this should wipe any left-over and restart from a fresh default configuration.
The point was that the flash left-over that Tasmota may use as boot Settings may be as bad as preventing Tasmota to start.

**Reply by jeroenst** [reply]
Ah ok, clear to me, thank you. So the ota software in the esp should erase settings before writing the tasmota image ideally. 

Could Fast Power Cycle Device Recovery work when tasmota is not starting after ota flash? Or are invalid boot settings preventing this?

**Reply by barbudor** [reply]
I would expect the recovery mechanisms to work ... as long as the code is correctly flashed

Note that I never dived into neither ArduinoOTA nor TasmotaOTA and their respective memory map, but if the OTA code is being erased by the new code while flashing, then it won't work.
I worked on OTA systems a long time ago where either the OTA code was on protected flash sector, or at least was copied to RAM. No idea how this done here.

**Reply by Jason2866** [reply]
Left overs can make device failing before Tasmota functions can kick in

**Reply by jeroenst** [reply]
Afaik arduino ota copies the image to the unused flash and moves it to the start of the flash after upgrading, so in theory it could wipe the first half of the flash without any problem.

Anyhow, I think I have learned enough, and I'm going to try things out with a test device.

Thank you for your time and effort!
